File: herencia-dto/AMarotoChivite-HerenciaDTO/AMaroto-Chivite-HerenciasDto-Java/src/main/java/Ficheros/HerenciasDto/controllers/ControllerQueries.java
package Ficheros.HerenciasDto.controllers;

import Ficheros.HerenciasDto.models.dto.PersonDTO;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ControllerQueries {

    private final List<PersonDTO> listOfStorageReadFile;
    private List<PersonDTO.Profesor> listProfesor;
    private List<PersonDTO.Alumno> listAlumno;

    public ControllerQueries(List<PersonDTO> listOfStorageReadFile) {
        this.listOfStorageReadFile = listOfStorageReadFile;
        initStorages();
    }

    private void initStorages() {
        this.listAlumno = listOfStorageReadFile.stream()
                .filter(personDTO -> personDTO instanceof PersonDTO.Alumno)
                .map(personDTO -> (PersonDTO.Alumno) personDTO)
                .collect(Collectors.toList());

        this.listProfesor = listOfStorageReadFile.stream()
                .filter(personDTO -> personDTO instanceof PersonDTO.Profesor)
                .map(personDTO -> (PersonDTO.Profesor) personDTO)
                .collect(Collectors.toList());
    }

    public PersonDTO.Profesor profesorMoreAncient() {
        if (!listProfesor.isEmpty()) {
            return Collections.max(listProfesor, Comparator.comparing(PersonDTO::getEdad));
        } else {
            return null;
        }
    }

    public PersonDTO.Alumno alumnoLessAncient() {
        return Collections.min(listAlumno, Comparator.comparing(PersonDTO::getEdad));
    }

    public double averageAgeAlumno() {
        return listAlumno.stream().mapToInt(PersonDTO::getEdad).average().orElse(Double.NaN);
    }

    public int averageLongitudeName() {
        return (int) listOfStorageReadFile.stream().mapToInt(personDTO -> personDTO.getName().length()).average().orElse(Double.NaN);
    }

    public Map<String, List<PersonDTO>> groupByType() {
        return listOfStorageReadFile.stream().collect(Collectors.groupingBy(personDTO -> personDTO.getClass().getSimpleName()));
    }
}

File: herencia-dto/AMarotoChivite-HerenciaDTO/AMarotoChivite-HerenciasDto-Kotlin/src/main/kotlin/Ficheros/HerenciasDto/controllers/ControllerQueries.kt
package Ficheros.HerenciasDto.controllers

import Ficheros.HerenciasDto.models.dto.PersonDTO

class ControllerQueries(private val listOfStorageReadFile: List<PersonDTO>) {

    private val listAlumno =
        listOfStorageReadFile.filter { it::class.simpleName == PersonDTO.Alumno::class.simpleName }
    private val listProfesor =
        listOfStorageReadFile.filter { it::class.simpleName == PersonDTO.Profesor::class.simpleName }

    fun profesorMoreAncient(): PersonDTO.Profesor? {
        if (listProfesor.isNotEmpty()) {
            return listProfesor.maxBy { it.edad } as PersonDTO.Profesor
        } else {
            return null
        }
    }

    fun alumnoLessAncient(): PersonDTO.Alumno {
        return listAlumno.minBy { it.edad } as PersonDTO.Alumno
    }

    fun averageAgeAlumno(): Double {
        return listAlumno.sumOf { it.edad.toDouble() } / listAlumno.size
    }

    fun averageLongitudeName(): Int {
        return listOfStorageReadFile.sumOf { it.name.length } / listAlumno.size
    }

    fun groupByType(): Map<String, List<PersonDTO>> {
        return listOfStorageReadFile
            .groupBy { it::class.simpleName.toString() }
            .mapValues { it ->
                it.value.map { it }
            }
    }
}